schema_registry.cc

```diff
@@ -274,39 +275,73 @@ global_schema_ptr::global_schema_ptr(global_schema_ptr&& o) noexcept {
+                    return e.frozen();
+                });
+            }
+            return ret;
+        };
+
+        schema_ptr registered_bs;
+        // the following code contains registry entry dereference of a foreign shard
+        // however, it is guarantied to succeed since we made sure in the constructor
+        // that _bs_schema and _ptr will have a registry on the foreign shard where this
+        // object originated so as long as this object lives the registry entries lives too
+        // and it is safe to reference them on foreign shards.
+        if (_base_schema) {
+            registered_bs = registered_schema(*_base_schema->registry_entry());
+            if (_base_schema->registry_entry()->is_synced()) {
+                registered_bs->registry_entry()->mark_synced();
+            }
         }
-        if (e.is_synced()) {
+        schema_ptr s = registered_schema(*_ptr->registry_entry());
+        if (s->is_view()) {
+            if (! s->view_info()->base_info()) {
+                // we know that registered_bs is valid here because we make sure of it in the constructors.
+                s->view_info()->set_base_info(s->view_info()->make_base_dependent_view_info(*registered_bs));
+            }
+        }
+        if (_ptr->registry_entry()->is_synced()) {
             s->registry_entry()->mark_synced();
         }
         return s;
     }
 }
 
 global_schema_ptr::global_schema_ptr(const schema_ptr& ptr)
-    : _ptr([&ptr]() {
-        // _ptr must always have an associated registry entry,
-        // if ptr doesn't, we need to load it into the registry.
-        schema_registry_entry* e = ptr->registry_entry();
-        if (e) {
-            return ptr;
+        : _cpu_of_origin(this_shard_id()) {
+    // _ptr must always have an associated registry entry,
+    // if ptr doesn't, we need to load it into the registry.
+    schema_registry_entry* e = ptr->registry_entry();
+    schema_ptr s;
+    if (e) {
+        s = ptr;
+    } else {
+        s = local_schema_registry().get_or_load(ptr->version(), [&ptr] (table_schema_version) {
+            return frozen_schema(ptr);
+        });
+    }
+    if (s->is_view()) {
+        if (s->view_info()->base_info()) {
+            _base_schema = global_schema_ptr(s->view_info()->base_info()->base_schema());
+        } else if (ptr->view_info()->base_info()) {
+            _base_schema = global_schema_ptr(ptr->view_info()->base_info()->base_schema());
+        } else {
+            on_internal_error(slogger,"Tried to build a global schema for view with an uninitialized base info");
         }
-        return local_schema_registry().get_or_load(ptr->version(), [&ptr] (table_schema_version) {
-                return frozen_schema(ptr);
-            });
-        }())
-    , _cpu_of_origin(this_shard_id())
-{ }
+        if (_base_schema) {
+            s->view_info()->set_base_info(s->view_info()->make_base_dependent_view_info(*_base_schema));
```

Comment:
We should do this only in the `if (ptr->view_info()->base_info())` case, in other cases it's not necessary because s already has base info attached.

Reply:
I can (and will) refine the condition here but it is not exactly what you wrote. The info is always attached the real
question is if this info has registry entry or not. A global view table which has a non registered schema as its base is useless.
It is true that if all other code paths are covered your condition and what I wrote above converges since a view schema is in the registry iff its base table is in the registry.
To be honest I prefer have here the additional (refined) safety check because I am not confident that all other paths into the registry conform to this invariant.